File: packaging/mcpb/kotlin/stackql-mcp/src/main/kotlin/io/stackql/mcp/Acquire.kt
package io.stackql.mcp

import java.nio.file.Files
import java.nio.file.Path

/**
 * Resolve a runnable stackql binary. Resolution order (first match wins):
 *
 *  1. `STACKQL_MCP_BIN` env: run that binary directly, no acquisition.
 *  2. Builder `binary` override: same.
 *  3. `STACKQL_MCP_BUNDLE` env: extract that local `.mcpb`. No pin check - an
 *     explicit override is operator intent and may be a custom build.
 *  4. Builder `bundlePath` override: same.
 *  5. Shared cache: an already-extracted binary for the pinned version.
 *  6. Verified download: fetch the pinned `.mcpb`, check its sha256 against the
 *     baked-in pin, extract, cache. Subsequent starts are offline.
 */
internal object Acquire {

    /** Inputs collected from the builder; env overrides are read here and win. */
    class Inputs(
        val binary: Path? = null,
        val bundlePath: Path? = null,
        val cacheRoot: Path = Cache.defaultRoot(),
        val version: String = Pins.STACKQL_VERSION,
    )

    fun resolveBinary(inputs: Inputs): Path {
        envPath(Cache.ENV_BIN)?.let { return existing(it, Cache.ENV_BIN) }
        inputs.binary?.let { return existing(it, "Builder.binary") }
        envPath(Cache.ENV_BUNDLE)?.let { return extractLocalBundle(it, Cache.ENV_BUNDLE, inputs) }
        inputs.bundlePath?.let { return extractLocalBundle(it, "Builder.bundlePath", inputs) }
        return sidecar(inputs)
    }

    private fun envPath(name: String): Path? =
        System.getenv(name)?.takeIf { it.isNotEmpty() }?.let { Path.of(it) }

    private fun existing(path: Path, what: String): Path {
        if (!Files.isRegularFile(path)) {
            throw StackqlMcpException("$what points at $path, which is not a file")
        }
        return path
    }

    /** Extract a caller-supplied `.mcpb` into a slot keyed by its content hash. */
    private fun extractLocalBundle(bundlePath: Path, what: String, inputs: Inputs): Path {
        if (!Files.isRegularFile(bundlePath)) {
            throw StackqlMcpException("$what points at $bundlePath, which is not a file")
        }
        val digest = Cache.sha256(bundlePath)
        val dest = inputs.cacheRoot.resolve("custom").resolve(digest.substring(0, 16))
        return BundleExtractor.extract(bundlePath, dest)
    }

    private fun sidecar(inputs: Inputs): Path {
        val platform = Platform.current()
        val dest = Cache.binaryDir(inputs.cacheRoot, inputs.version, platform)
        BundleExtractor.cachedBinary(dest)?.let { return it }

        val pin = Pins.pinFor(platform)
        val url = Pins.bundleUrl(platform, inputs.version)
        val mcpb = inputs.cacheRoot.resolve(inputs.version).resolve(platform.bundleName)
        System.err.println("stackql-mcp: downloading $url (first run, cached at $dest)")
        Download.verified(url, pin, mcpb)
        val binary = BundleExtractor.extract(mcpb, dest)
        // The extracted dir is the cache; drop the archive to halve disk use.
        Files.deleteIfExists(mcpb)
        return binary
    }
}

File: packaging/mcpb/cargo/src/acquire.rs
//! Binary acquisition: resolve a runnable stackql binary from (in order)
//! env overrides, builder overrides, vendored bytes, or sidecar download.

use std::fs;
use std::path::{Path, PathBuf};

use crate::bundle;
use crate::cache;
use crate::error::{Error, Result};
#[cfg(feature = "sidecar")]
use crate::platform::Platform;

/// Acquisition inputs collected from the builder. Env overrides are read at
/// resolve time and take precedence over everything here.
#[derive(Default)]
pub struct Acquisition {
    /// Run this binary directly, skipping bundles entirely.
    pub binary: Option<PathBuf>,
    /// Extract this local .mcpb instead of downloading.
    pub bundle_path: Option<PathBuf>,
    /// Embedded .mcpb bytes (vendored feature).
    #[cfg(feature = "vendored")]
    pub bundle_bytes: Option<&'static [u8]>,
}

/// Resolve the server binary, acquiring it if needed. Blocking; call from a
/// blocking context (`start()` wraps this in `spawn_blocking`).
pub fn resolve_binary(acq: &Acquisition) -> Result<PathBuf> {
    // 1. Env binary override: run it as-is.
    if let Some(bin) = std::env::var_os(cache::ENV_BIN).filter(|v| !v.is_empty()) {
        return existing(PathBuf::from(bin), cache::ENV_BIN);
    }
    // 2. Builder binary override.
    if let Some(bin) = &acq.binary {
        return existing(bin.clone(), "Builder::binary");
    }
    // 3. Env bundle override: extract a local .mcpb. No pin check - the
    //    override is explicit operator intent and may be a custom build.
    if let Some(bundle_path) = std::env::var_os(cache::ENV_BUNDLE).filter(|v| !v.is_empty()) {
        return extract_local_bundle(&PathBuf::from(bundle_path), cache::ENV_BUNDLE);
    }
    // 4. Builder bundle override.
    if let Some(bundle_path) = &acq.bundle_path {
        return extract_local_bundle(bundle_path, "Builder::bundle_path");
    }
    // 5. Vendored bytes embedded by the caller.
    #[cfg(feature = "vendored")]
    if let Some(bytes) = acq.bundle_bytes {
        return extract_vendored(bytes);
    }
    // 6. Sidecar: shared cache, then verified download.
    #[cfg(feature = "sidecar")]
    {
        return sidecar();
    }
    #[allow(unreachable_code)]
    Err(Error::Bundle(
        "no binary source: enable the sidecar feature, embed a bundle with the vendored \
         feature, or set STACKQL_MCP_BIN / STACKQL_MCP_BUNDLE"
            .into(),
    ))
}

fn existing(path: PathBuf, what: &'static str) -> Result<PathBuf> {
    if path.is_file() {
        Ok(path)
    } else {
        Err(Error::OverrideNotFound { what, path })
    }
}

/// Extract a caller-supplied .mcpb into a cache slot keyed by its content
/// hash, so different bundles never collide.
fn extract_local_bundle(bundle_path: &Path, what: &'static str) -> Result<PathBuf> {
    if !bundle_path.is_file() {
        return Err(Error::OverrideNotFound {
            what,
            path: bundle_path.to_path_buf(),
        });
    }
    let digest = crate::download::sha256_file(bundle_path)?;
    let dest = cache::bin_cache_root()?.join("custom").join(&digest[..16]);
    if let Some(binary) = bundle::cached_binary(&dest) {
        return Ok(binary);
    }
    let file = fs::File::open(bundle_path)?;
    bundle::extract_bundle(file, &dest)
}

#[cfg(feature = "vendored")]
fn extract_vendored(bytes: &'static [u8]) -> Result<PathBuf> {
    use sha2::{Digest, Sha256};
    let digest = Sha256::digest(bytes);
    let mut key = String::with_capacity(16);
    for b in &digest[..8] {
        key.push_str(&format!("{b:02x}"));
    }
    let dest = cache::bin_cache_root()?.join("vendored").join(key);
    if let Some(binary) = bundle::cached_binary(&dest) {
        return Ok(binary);
    }
    bundle::extract_bundle(std::io::Cursor::new(bytes), &dest)
}

#[cfg(feature = "sidecar")]
fn sidecar() -> Result<PathBuf> {
    let platform = Platform::detect()?;
    let pin = crate::pins::pin_for(platform)?;
    let dest = cache::bundle_cache_dir(crate::pins::STACKQL_VERSION, platform.key())?;
    if let Some(binary) = bundle::cached_binary(&dest) {
        return Ok(binary);
    }

    let url = crate::pins::bundle_url(pin);
    let mcpb = cache::bin_cache_root()?
        .join(crate::pins::STACKQL_VERSION)
        .join(pin.bundle_name);
    eprintln!(
        "stackql-mcp: downloading {} (first run, cached at {})",
        url,
        dest.display()
    );
    crate::download::download_verified(&url, pin.sha256, &mcpb)?;
    let file = fs::File::open(&mcpb)?;
    let binary = bundle::extract_bundle(file, &dest)?;
    // The extracted dir is the cache; drop the archive to halve disk use.
    let _ = fs::remove_file(&mcpb);
    Ok(binary)
}
